### ISSUE TITLE
Check for multiple statements in prepare

### ISSUE THREAD

**Issue by thomcc** [rusqlite/rusqlite#1147]
Consider checking by default if someone does something like `conn.prepare("<stmt 1>; <stmt 2>")`. This will silently only run stmt 1, which is probably not what the user wants.

Right now, I believe we have some tail checking here, but I think it's only under extra-check feature. It also might be disabled then too. I think we should probably check this always, since I don't see a case where anybody wouldn't want this detected.

**Comment by gwenn** [comment]
You may have false positive with some trailing comments "SELECT 1; /*comments*/;"

**Comment by thomcc** [comment]
Yeah. I think that's probably fine.

**Comment by jvasile** [comment]
I ran into this issue in my own code and worked up a crate to quickly split strings containing multiple sqlite statements and also to tell me when I'm dealing with multiple sql statements.  It's standalone functionality, so I put it in [its own crate](https://github.com/jvasile/sql_split/), but I could easily turn it into a PR.

`sql_split` is O(n) to split.  It also has a short-circuiting func, `is_multiple`, that just tells you if you're dealing with a multi-statement string.

If this is useful to rusqlite, I'd be happy to turn it into a PR.  My thinking is that we shouldn't just split every statement that comes in but rather might make available versions of functions that DTRT with multiple statement strings.  I'd hate to pay the O(n) penalty on a large number of sql statements if I *know* they're single-statement strings.  In that case, some documentation tweaks might be helpful to guide devs to the version they want.

Alternatively, one might be able to avoid parsing and abuse the heck out of [SQLite's complete function](https://sqlite.org/c3ref/complete.html), but I can't think of a way to do that efficiently.

**Comment by gwenn** [comment]
- [ ] https://sqlite.org/lang_createtrigger.html and statements between `BEGIN` and `END` => you cannot split easily
- [ ] https://sqlite.org/lang_createvtab.html and module arguments which can be anything (`CREATE VIRTUAL TABLE my_csv USING csv(filename=..., delimiter=;);`)

**Comment by thomcc** [comment]
I think we should take the current approach, where we used the tail returned by SQLite. This has downsides, but is much less error-prone than trying to parse it ourselves IMO.

**Comment by jvasile** [comment]
Yes, that's right.  I thought it could be done simply, but after chasing down a bunch of edge cases and handling virtual tables, I think this "simple" bit of code will quickly get hairier than its worth.

**Comment by d-bucur** [comment]
Would also be great if there was an equivalent method for multiple statements, like there is for `execute_batch`. The only way to do that currently is to manually format the string and send it to execute_batch

**Comment by gwenn** [comment]
We should be able to check at compile time that only a single statement is provided with [sqlite3-parser](https://crates.io/crates/sqlite3-parser) if the SQL is a static string.

**Comment by gwenn** [comment]
There is a new `single!` macro [here](https://github.com/rusqlite/rusqlite/pull/1455/commits/97a13e21a45cc1be15142b45b5e50dd8a438b77f#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759).
But it cannot be used inside `rusqlite` existing method.
For example, it cannot work if put inside [Connection::execute](https://github.com/rusqlite/rusqlite/blob/a96f61b17fa96c58cb5c2f9c15bf11311e129707/src/lib.rs#L621) method.
I can work only if used like this:
```rust
let changes = conn.execute(single!("DELETE FROM my_table WHERE col = ?"), ...)?;
```
which is not user-friendly...
Do you know how to make it nicer (without duplicating `rusqlite` public API) ?

**Comment by ijackson** [comment]
I just spent a little while figuring out why half my schema wasn't being run.  This feature would have saved me.  The current behaviour is something of a footgun.

**Comment by ijackson** [comment]
My reading of #1679 is that it does not affect `.prepare()`.  Am I wrong about that?

If I'm right, I think this ticket should be reopened.

**Comment by gwenn** [comment]
> My reading of [#1679](https://github.com/rusqlite/rusqlite/pull/1679) is that it does not affect `.prepare()`. Am I wrong about that?
> 
> If I'm right, I think this ticket should be reopened.

You are right => #1680

**Comment by ijackson** [comment]
Thanks :-).
